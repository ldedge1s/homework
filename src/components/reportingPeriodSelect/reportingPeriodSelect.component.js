import InputLabel from "@mui/material/InputLabel";
import Select from "@mui/material/Select";
import React, {useMemo, useCallback} from "react";
import MenuItem from "@mui/material/MenuItem";
import {MONTH_LABELS} from "../../constants/monthLabels";
import PropTypes from "prop-types";
import {styled} from "@mui/material";

const reportingPeriod = 3;
const reportingPeriodEndOffset = reportingPeriod - 1;
const maxReportingStartMonthIdx = 11 - reportingPeriodEndOffset;

const selectValueOption = {
    value: 'select',
    label: 'Select'
};

const StyledSelect = styled(Select)({
    width: 220
});

const ReportingPeriodSelect = ({value, onPeriodChange}) => {
    const optionsData = useMemo(() => {
        const newOptionsData = [selectValueOption];
        for (let i = 0; i <= maxReportingStartMonthIdx; i++) {
            const value = {
                start: i,
                end: i + reportingPeriodEndOffset,
            };

            newOptionsData.push({
                value,
                label: `${MONTH_LABELS[value.start]} - ${MONTH_LABELS[value.end]}`,
            });
        }
        return newOptionsData;
    }, []);

    const handleChange = useCallback((evt) => {
        if (onPeriodChange) {
            const newValue = evt?.target?.value;
            if (newValue === selectValueOption.value) {
                onPeriodChange(undefined);
            } else {
                onPeriodChange(newValue);
            }
        }
    }, [onPeriodChange])

    if (!value) {
        value = selectValueOption.value;
    }

    return (
        <>
            <InputLabel id="demo-simple-select-label">Reporting Period</InputLabel>
            <StyledSelect
                labelId="reporting-period"
                label="Reporting Period"
                value={value}
                onChange={handleChange}
            >
                {optionsData.map((data, idx) => (
                    <MenuItem value={data.value} key={idx}>
                        {data.label}
                    </MenuItem>
                ))}
            </StyledSelect>
        </>
    );
}

ReportingPeriodSelect.propTypes = {
    value: PropTypes.shape({
        value: PropTypes.string,
        label: PropTypes.string
    }),
    onPeriodChange: PropTypes.func
};

ReportingPeriodSelect.defaultProps = {
    onPeriodChange: () => {
    },
};

export {
    ReportingPeriodSelect
}
