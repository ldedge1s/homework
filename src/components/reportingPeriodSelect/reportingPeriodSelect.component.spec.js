/**
 * @jest-environment jsdom
 */

import React from "react";
import {render, fireEvent, within} from '@testing-library/react';
import '@testing-library/jest-dom';
import {ReportingPeriodSelect} from "./reportingPeriodSelect.component";

describe('reportingPeriodSelect component', () => {
    it('renders properly', () => {
        //given
        const {container} = render(<ReportingPeriodSelect/>);

        //then
        expect(container).toMatchSnapshot();
    });

    it('renders proper options', () => {
        //given
        const {getByRole} = render(<ReportingPeriodSelect/>);

        //when
        fireEvent.mouseDown(getByRole('button'));

        //then
        const listbox = within(getByRole('listbox'));
        expect(listbox.getAllByRole('option')).toMatchSnapshot();
    });

    it('calls onPeriodChange properly', () => {
        //given
        const handleChange = jest.fn();
        const {getByRole} = render(<ReportingPeriodSelect onPeriodChange={handleChange}/>);

        //when
        fireEvent.mouseDown(getByRole('button'));
        const listbox = within(getByRole('listbox'));

        expect(handleChange).not.toBeCalled();
        fireEvent.click(listbox.getAllByRole('option')[2]);

        //then
        expect(handleChange).toHaveBeenCalledTimes(1);
        expect(handleChange).toHaveBeenNthCalledWith(1, {start: 1, end: 3});
    });
});


