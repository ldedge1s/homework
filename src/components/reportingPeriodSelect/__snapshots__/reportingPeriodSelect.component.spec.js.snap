// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reportingPeriodSelect component renders proper options 1`] = `
Array [
  <li
    aria-selected="true"
    class="MuiMenuItem-root MuiMenuItem-gutters Mui-selected MuiButtonBase-root Mui-focusVisible css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="select"
    role="option"
    tabindex="0"
  >
    Select
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    January - March
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    February - April
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    March - May
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    April - June
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    May - July
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    June - August
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    July - September
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    August - October
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    September - November
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
  <li
    aria-selected="false"
    class="MuiMenuItem-root MuiMenuItem-gutters MuiButtonBase-root css-kk1bwy-MuiButtonBase-root-MuiMenuItem-root"
    data-value="[object Object]"
    role="option"
    tabindex="-1"
  >
    October - December
    <span
      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
    />
  </li>,
]
`;

exports[`reportingPeriodSelect component renders properly 1`] = `
<div>
  <label
    class="MuiInputLabel-root MuiInputLabel-animated MuiFormLabel-root MuiFormLabel-colorPrimary css-9npbnl-MuiFormLabel-root-MuiInputLabel-root"
    id="demo-simple-select-label"
  >
    Reporting Period
  </label>
  <div
    class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary css-46xaai-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
  >
    <div
      aria-expanded="false"
      aria-haspopup="listbox"
      aria-labelledby="reporting-period"
      class="MuiSelect-select MuiSelect-outlined MuiOutlinedInput-input MuiInputBase-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
      role="button"
      tabindex="0"
    >
      Select
    </div>
    <input
      aria-hidden="true"
      class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
      tabindex="-1"
      value="select"
    />
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
      data-testid="ArrowDropDownIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M7 10l5 5 5-5z"
      />
    </svg>
    <fieldset
      aria-hidden="true"
      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
    >
      <legend
        class="css-1ftyaf0"
      >
        <span>
          Reporting Period
        </span>
      </legend>
    </fieldset>
  </div>
</div>
`;
